src/plugins/intel_gpu/src/graph/graph_optimizer/reshape_transfer.cpp

```diff
@@ -0,0 +1,143 @@
+// Copyright (C) 2025 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include <algorithm>
+
+#include "pass_manager.h"
+#include "permute_inst.h"
+#include "program_helpers.h"
+#include "reorder_inst.h"
+#include "reshape_inst.h"
+
+using namespace cldnn;
+
+void reshape_transfer::run(program& p) {
+    // (reorder) + reshape + transpose
+    // sink reshape for further possible optimization
+    auto is_suitable_permute = [](cldnn::program_node* node) {
+        return node->get_users().size() == 1 && node->is_dynamic() == false &&
+               node->get_output_layout().get_rank() == 4;
+    };
+
+    auto is_suitable_reshape = [](cldnn::program_node* node) -> bool {
+        if (node->get_users().size() != 1 || node->is_dynamic())
+            return false;
+        auto& input_lay = node->get_input_layout(0);
+        auto& output_lay = node->get_output_layout();
+        if (input_lay.compatible(output_lay))
+            return true;
+        return false;
+    };
+
+    std::function<bool(cldnn::program_node*)> is_suitable_parent;
+    is_suitable_parent = [&is_suitable_parent](const cldnn::program_node* node) -> bool {
+        if (node->get_users().size() != 1 || node->is_dynamic())
+            return false;
+        if (node->is_type<convolution>())
+            return true;
+        for (size_t idx = 0; idx < node->get_dependencies().size(); idx++) {
+            auto& input = node->get_dependency(idx);
+            if (!input.is_in_data_flow() || input.is_constant())
+                continue;
+            if (node->is_type<reorder>()) {
+                return is_suitable_parent(&input);
+            } else if (node->is_type<eltwise>() && node->get_dependency(1).is_constant()) {
+                return is_suitable_parent(&input);
+            } else if (node->is_type<activation>()) {
+                return is_suitable_parent(&input);
+            }
+            return false;
+        }
+        return false;
+    };
+
+    auto update_order = [](std::vector<uint16_t> original_order, cldnn::program_node* reshape) {
+        if (!reshape)
+            return original_order;
+        // Example. For this sequence, there is Reshape node which merges 2 consecutive dims into one
+        // order must be updated like permute is done before reshape
+        // [1,3,4,6] -> Reshape[1,3,24,1]-> permute(0,2,1) -> [1,24,3,1]
+        // updated order must be (0,2,3,1):
+        // dim with index=2 is split into 2 parts: 2 and 3
+        const auto& reshape_in_shape = reshape->get_input_layout().get_dims();
+        const auto& reshape_out_shape = reshape->get_output_layout().get_dims();
+        auto transformed_order = original_order;
+        ov::Shape new_shape(transformed_order.size());
+        const uint16_t merge_dim_idx = [&]() {
+            for (uint16_t i = 0; i < reshape_in_shape.size(); ++i) {
+                if (reshape_in_shape[i] != reshape_out_shape[i])
+                    return i;
```

Comment:
what if
reshape[1, 6, 1, 1] => [1, 2, 3, 1]
permute order[2, 3] ?

Reply:
Also what if input/output ranks are different?

Reply:
> Also what if input/output ranks are different?

yeah..add more strict check for reshape node

Reply:
> what if reshape[1, 6, 1, 1] => [1, 2, 3, 1] permute order[2, 3] ?

do you mean permute order (0,2,3,1)?
so we have [1,6,1,1] -> reshape [1,2,3,1] -> permute (0,2,3,1) [1,3,1,2]
we have updated permute order as (0,3,1,2) which is not rotating feature dim to back to be taken as inner-most axis, thus it will fail the pattern and not sink reshape in this case